### ISSUE TITLE
P4-16 table with multiple actions and no default_action causes compiler bug

### ISSUE THREAD

**Issue by hesingh** [p4lang/p4c#1166]
In P4-16 code being compiled with p4c compiler, if a table has multiple actions and the table has not defined a default_action, the following compiler bug is emitted when p4c compiles the P4 program.  The fix should be to have p4c catch that the table has no default_action but the table has multiple actions.

In file: /home/hemant/Downloads/int/p4c/ir/type.def:126
[31mCompiler Bug[0m: Only 0 parameters; index #0 requested

The table example is included below.

table xpFdbTableDa {
      size = 262144;
        key = {
         fdb_kd.Da         : exact;
         fdb_kd.bd         : exact;
         fdb_kd.key_type   : exact;
        }
        actions = {processFdbTblDa; processFdbTblDaMiss;}
      //  default_action = processFdbTblDaMiss;
        implementation = hash_table(6);
}

**Comment by jafingerhut** [comment]
I tried to reproduce the crash with the latest p4c as of today, and didn't find a way to do it using 'p4test' or 'p4c-bm2-ss' on the attached full P4_16 program, which I know comments out the 'implementation' line, and has different keys for the table, but it does have the 2 actions, and a commented-out default_action.

If you have a full program that crashes the latest version of p4c, it would be helpful to attach that full program, along with the command line you used that crashed, and the version of the compiler you used.
[issue1166.p4.txt](https://github.com/p4lang/p4c/files/1746693/issue1166.p4.txt)

**Comment by hesingh** [comment]
I am using a p4c from around September-end 2017.  Let me re-test with a latest p4c and get back.

**Comment by hesingh** [comment]
Hi Andy,

I downloaded the latest p4c and I also cannot reproduce the compiler crash.   However, before we close this issue, I am wondering if the compiler should emit an error if a table has multiple actions defined, but no default_action defined?  thanks.

**Comment by jafingerhut** [comment]
I do not believe the compiler should emit an error for such a table.

The P4_16 language specification, end of Section 12.2.1.3 "Default Action", says this:

"If a table does not specify the default_action property and no entry matches a given packet, then the table does not affect the packet and processing continues according to the imperative control flow of the program."

https://p4.org/p4-spec/docs/P4-16-v1.0.0-spec.html

So the behavior is well defined for tables that have no default_action defined, regardless of how many actions they have.

**Comment by hesingh** [comment]
Sorry, I sent an email to you and Mihai and then saw your reply to this Issue.   I think we can Close this issue once Mihai confirms the intended behavior

**Comment by hesingh** [comment]
I saw this text in the P4-16 specification where the text includes a "may" and thus I wasn't sure what the compiler does.  I tested the latest p4c and see the compiler adds a default_action.  Thanks to Mihai to confirm that the compiler will adhere to this behavior. 

[The compiler may set the default_action to NoAction (and also insert it into the list of actions) for tables that do not define the default_action property. This is consistent with the semantics given in Section 12.2.1.3. In this document, we assume that that this transformation has been performed, so that all tables have a default_action property.]
